[FLINK-22881][runtime] Revert toggling IDLE/ACTIVE on records in IDLE state

dawidwys opened the pull request:
## What is the purpose of the change

We revert the toggling between IDLE and ACTIVE status if a record is emitted in an IDLE state. If records are emitted in an IDLE state the toggling incur a lot of performance penalty.

## Verifying this change

All existing tests pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:
```diff
@@ -110,12 +108,14 @@ public void collect(StreamRecord<OUT> record) {
-        try (AutoCloseable ignored = announcedStatus.ensureActive(this::writeStreamStatus)) {
-            watermarkGauge.setCurrentWatermark(mark.getTimestamp());
-            serializationDelegate.setInstance(mark);
+        if (announcedStatus.isIdle()) {
+            return;
+        }
```
This is as it it in 1.13 - emitting watermarks do not switch the status to active. But that's something that we might want to change in the future?

dawidwys replied:
Yes, in 1.13 we ignore/swallow watermarks emitted in the `IDLE`state. There we check the status via `StreamStatusProvider` though.

dawidwys replied:
The comment about tests reminded me that I dropped a test for it along with the behaviour in outputs.

pnowojski reviewed (commented):
Thanks for the quick fix. I think the change looks good to me (% one question), except of:

Is there just a single test modified as a result of this change? I would expect a bit more.

3 methods of `RecordWriterOutput` were modified
2 in both `ChainingOutput` and `BroadcastingOutputCollector`.

It looks like in unit tests only one of those 7 places is tested?

dawidwys commented:
There is one additional comprehensive test that I forgot to revert. I removed it because we removed the behaviour for swallowing watermarks in IDLE channels.

pnowojski commented:
Aren't we still missing a test coverage for broadcasting?

dawidwys commented:
For documentation purposes:
* `RecordWriterOutput` is tested in `OneInputStreamTaskTest#testWatermarkAndStreamStatusForwarding`
* `ChainingOutput` is tested in `OneInputStreamTaskTest#testWatermarksNotForwardedWithinChainWhenIdle`

When working on the test for `BroadcastingOutputCollector` I realized bookkeeping the `announcedStatus` there is unnecessary. It will always forward its output to either the `ChainingOutput` or the `RecordWriterOutput` which have that logic already. There is no need to double the check the status.

pnowojski reviewed (approved):
LGTM % test failure - it looks unrelated, but it hasn't been reported before?
